File: Week 28/java/minimax.java
/*
 * WEEK 28 - JAVA ADVANCED TOPICS
 * Topic: Minimax (Zero-Sum Game Tree Search)
 * File: minimax.java
 *
 * CONCEPT:
 *     For a two-player, perfect-information, zero-sum game, minimax assigns
 *     a value v(s) to every state s representing the score the *maximising*
 *     player can guarantee against an optimally-playing *minimising*
 *     opponent. The recurrence:
 *
 *         v(s) = utility(s)                           if s is terminal
 *              = max over moves of v(child(s))        if Max to move
 *              = min over moves of v(child(s))        if Min to move
 *
 *     The depth-first traversal of the full game tree computes v(root).
 *
 * KEY POINTS:
 *     - Evaluates every reachable leaf -> O(b^d). Use alpha-beta to prune.
 *     - Memoisation (transposition table) helps when the same state is
 *       reachable by multiple move orderings.
 *     - For non-finite games use a depth limit + heuristic evaluator.
 *
 * ALGORITHM / APPROACH:
 *     minimax(state, maxTurn):
 *         if terminal(state): return utility(state)
 *         if maxTurn: return max over children of minimax(child, false)
 *         else:       return min over children of minimax(child, true)
 *
 * DRY RUN / EXAMPLE:
 *     Empty 3x3 tic-tac-toe with X to move -> minimax returns 0 (draw with
 *     optimal play). Best opening grabs the centre (or any corner).
 *
 * COMPLEXITY:
 *     Time:  O(b^d) without pruning; here d <= 9 for TTT -> ~360k positions.
 *     Space: O(d) recursion; O(|states|) when memoised.
 */

// snake_case filename is fine; class Minimax is package-private.

import java.util.HashMap;
import java.util.Map;

class Minimax {

    static final int[][] LINES = {
        {0, 1, 2}, {3, 4, 5}, {6, 7, 8},
        {0, 3, 6}, {1, 4, 7}, {2, 5, 8},
        {0, 4, 8}, {2, 4, 6}
    };

    static char winner(char[] board) {
        for (int[] line : LINES) {
            char a = board[line[0]];
            if (a != '.' && a == board[line[1]] && a == board[line[2]]) return a;
        }
        return 0;
    }

    static boolean isFull(char[] board) {
        for (char c : board) if (c == '.') return false;
        return true;
    }

    static int utility(char[] board) {
        char w = winner(board);
        if (w == 'X') return +1;
        if (w == 'O') return -1;
        return 0;
    }

    // Memoisation cache keyed by board string and side to move.
    static Map<String, Integer> memo = new HashMap<>();

    static int minimax(char[] board, boolean maxTurn) {
        char w = winner(board);
        if (w != 0 || isFull(board)) return utility(board);

        String key = new String(board) + (maxTurn ? 'X' : 'O');
        Integer cached = memo.get(key);
        if (cached != null) return cached;

        int best = maxTurn ? -2 : +2;
        char mark = maxTurn ? 'X' : 'O';
        for (int i = 0; i < 9; i++) {
            if (board[i] != '.') continue;
            board[i] = mark;
            int v = minimax(board, !maxTurn);
            board[i] = '.';
            if (maxTurn) best = Math.max(best, v);
            else         best = Math.min(best, v);
        }
        memo.put(key, best);
        return best;
    }

    /** Returns the optimal move index (0..8) for the side to move. */
    static int bestMove(char[] board, boolean maxTurn) {
        int bestIdx = -1;
        int bestVal = maxTurn ? -3 : +3;
        char mark = maxTurn ? 'X' : 'O';
        for (int i = 0; i < 9; i++) {
            if (board[i] != '.') continue;
            board[i] = mark;
            int v = minimax(board, !maxTurn);
            board[i] = '.';
            if (maxTurn && v > bestVal) { bestVal = v; bestIdx = i; }
            if (!maxTurn && v < bestVal) { bestVal = v; bestIdx = i; }
        }
        return bestIdx;
    }

    public static void main(String[] args) {
        char[] empty = "........." .toCharArray();
        System.out.println("Minimax value of empty TTT (X to move): " + minimax(empty, true));
        System.out.println("Best opening move for X: cell " + bestMove(empty, true));

        // X already in a corner -> minimax value still 0 against optimal O.
        char[] board = "X........".toCharArray();
        System.out.println("After X corner, minimax value: " + minimax(board, false));
    }
}

/*
 * NOTES
 * -----
 * Differences from the Python implementation in minimax.py:
 *   - Python uses tuples + @lru_cache; Java memoises via a HashMap keyed by
 *     the board string + side-to-move character.
 *   - We mutate a single char[] board in/out, which is faster than rebuilding
 *     immutable tuples on every recursive call.
 *   - The companion game_theory.java does NOT cover minimax; this file fills
 *     the gap, mirroring the Python and C++ splits.
 */

File: Week 28/rust/minimax.rs
// WEEK 28 - RUST ADVANCED TOPICS
// Topic: Minimax (Zero-Sum Game Tree Search)
// File: minimax.rs
//
// CONCEPT:
//   Recursively evaluate the value of every game-tree node where Max picks
//   the maximum of children and Min picks the minimum. Returns the game's
//   value with optimal play.
//
// KEY POINTS:
//   - O(b^d). Memoise hashable states for repeated subtrees.
//   - For non-finite games combine with depth limits + heuristic eval.
//
// ALGORITHM / APPROACH:
//   minimax(s, max_turn):
//     terminal -> utility(s)
//     else (max | min) over moves of minimax(child, !max_turn)
//
// RUST-SPECIFIC NOTES:
//   - Board as [char; 9] (Copy, Hash).
//   - HashMap<[char;9], i32> memoises.
//   - 'X' = +1, 'O' = -1.
//
// DRY RUN / EXAMPLE:
//   Empty TTT board, X to move -> value 0 (draw with optimal play).
//
// COMPLEXITY:
//   Time O(b^d); Space O(d) recursion + O(|states|) memo.

use std::collections::HashMap;

type Board = [char; 9];

const LINES: [[usize; 3]; 8] = [
    [0,1,2],[3,4,5],[6,7,8],[0,3,6],[1,4,7],[2,5,8],[0,4,8],[2,4,6]
];

fn winner(b: &Board) -> char {
    for l in LINES.iter() {
        if b[l[0]] != '.' && b[l[0]] == b[l[1]] && b[l[1]] == b[l[2]] {
            return b[l[0]];
        }
    }
    '.'
}

fn is_full(b: &Board) -> bool { !b.iter().any(|&c| c == '.') }

fn utility(b: &Board) -> i32 {
    match winner(b) { 'X' => 1, 'O' => -1, _ => 0 }
}

fn minimax(b: Board, max_turn: bool, memo: &mut HashMap<Board, i32>) -> i32 {
    if winner(&b) != '.' || is_full(&b) { return utility(&b); }
    if let Some(&v) = memo.get(&b) { return v; }
    let mut best = if max_turn { -2 } else { 2 };
    for i in 0..9 {
        if b[i] != '.' { continue; }
        let mut c = b;
        c[i] = if max_turn { 'X' } else { 'O' };
        let v = minimax(c, !max_turn, memo);
        if max_turn { if v > best { best = v; } } else { if v < best { best = v; } }
    }
    memo.insert(b, best);
    best
}

fn best_move(b: Board, max_turn: bool, memo: &mut HashMap<Board, i32>) -> usize {
    let mut idx = usize::MAX;
    let mut bv = if max_turn { -3 } else { 3 };
    for i in 0..9 {
        if b[i] != '.' { continue; }
        let mut c = b;
        c[i] = if max_turn { 'X' } else { 'O' };
        let v = minimax(c, !max_turn, memo);
        if (max_turn && v > bv) || (!max_turn && v < bv) {
            bv = v;
            idx = i;
        }
    }
    idx
}

fn main() {
    let mut memo: HashMap<Board, i32> = HashMap::new();
    let empty: Board = ['.'; 9];
    println!("Minimax value of empty TTT: {}", minimax(empty, true, &mut memo));
    println!("Best opening move for X: cell {}", best_move(empty, true, &mut memo));
}

// NOTES
// -----
// Differences from Java:
//   * Java's game_theory.java does not include minimax; we add it.
//   * [char; 9] is a Copy + Hash type, perfect for fast memo keys.
//   * Explicit HashMap memo is more straightforward in Rust than wrapping
//     a mutable closure.
